src/pure/List.mo

```diff
@@ -58,10 +58,12 @@ module {
   /// Runtime: O(size)
   ///
   /// Space: O(1)
-  public func size<T>(list : List<T>) : Nat = switch list {
-    case null 0;
-    case (?(_, t)) 1 + size t
-  };
+  public func size<T>(list : List<T>) : Nat = (
+    func go(n : Nat, list : List<T>) : Nat = switch list {
+      case (?(_, t)) go(n + 1, t);
+      case null n
+    }
+  )(0, list);
```

Comment:
```suggestion
  public func size<T>(list : List<T>) : Nat = {
     var size = 0;
     var tmp = list;
     loop { 
        switch tmp {
          case (?(_, tl)) {
             size += 1; 
             tmp := tl;
          };
          case null  return size
        }
   }
```
Is probably cheaper (no call to a helper).

Reply:
Yep, I was hoping that the compiler could create the same (or better) code by TCO. I.e. not sure if `var` allocates in this example (it shouldn't).

Reply:
I did the same transformation in `repeat`, though.

Reply:
reverse needs it too. TCO won't inline the auxialliary function so ....

Reply:
I actually test `reverse` for stack overflow and I haven't seen any 😖

Reply:
But for TCO one only have to turn the body of the helper into a loop, no need to inline it into the wrapper. Or am I mistaken?

Reply:
Of course the interpreter won't do TCO, and if we want constant stack there we'll have to manually TCO anyway...

Reply:
> I actually test `reverse` for stack overflow and I haven't seen any 😖 With depth 100_000!

If you test in the intrepeter you won't see because the interpreter is in CPS... On wasm, you should see

Reply:
I test on Wasm, and don't see it :-)

Reply:
Sorry, you reverse is tail recursive, but it allocates and calls a helper that is probably slower than just using a loop and a temp, without the additional function call.